File: Sources/Analytics/EventQueue/Helpers/Timestamp.m
//
//  Timestamp.m
//  PaltaLibAnalytics
//
//  Created by Vyacheslav Beltyukov on 20/07/2022.
//

#import <Foundation/Foundation.h>

@interface PBTimeKeeperLauncher : NSObject

@end

@implementation PBTimeKeeperLauncher

+ (void)load {
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wundeclared-selector"
    Class timeLoader = [[NSBundle bundleForClass:self] classNamed:@"PBTimeKeeper"];
    [timeLoader performSelector:@selector(recordTime)];
#pragma GCC diagnostic pop
}

@end

File: Sources/Analytics/EventQueue/Helpers/Timestamp.swift
//
//  Timestamp.swift
//  PaltaLibAnalytics
//
//  Created by Vyacheslav Beltyukov on 07/06/2022.
//

import Foundation

#if DEBUG
var mockedTimestamp: Int?
#endif

func currentTimestamp() -> Int {
    #if DEBUG
    return mockedTimestamp ?? realCurrentTimestamp()
    #else
    return realCurrentTimestamp()
    #endif
}

private var appStartTimestamp: Int!
private var appStartClock: Int!

private func realCurrentTimestamp() -> Int {
    appStartTimestamp + getClock() - appStartClock
}

private func getClock() -> Int {
    let nanosecs = clock_gettime_nsec_np(CLOCK_MONOTONIC_RAW)
    return Int(nanosecs / 1_000_000)
}

@objc(PBTimeKeeper)
private class TimeKeeper: NSObject {
    @objc
    class func recordTime() {
        appStartTimestamp = Int((Date().timeIntervalSince1970 * 1000).rounded())
        appStartClock = getClock()
    }
}
